src/Payments.sol

```diff
@@ -0,0 +1,700 @@
+        );
+
+        // Calculate the change in base lockup
+        uint256 oldLockup = rail.lockupFixed +
+            (rail.paymentRate * rail.lockupPeriod);
+        uint256 newLockup = lockupFixed + (rail.paymentRate * period);
+
+        // assert that operator allowance is respected
+        require(
+            newLockup <= oldLockup + approval.lockupAllowance,
+            "exceeds operator lockup allowance"
+        );
+        approval.lockupAllowance =
+            approval.lockupAllowance +
+            oldLockup -
+            newLockup;
+
+        // Update payer's lockup
+        require(
+            payer.lockupCurrent >= oldLockup,
+            "payer's current lockup cannot be less than old lockup"
+        );
+        payer.lockupCurrent = payer.lockupCurrent - oldLockup + newLockup;
+
+        // Update rail lockup parameters
+        rail.lockupPeriod = period;
+        rail.lockupFixed = lockupFixed;
+
+        require(
+            payer.lockupCurrent <= payer.funds,
+            "payer's current lockup cannot be greater than their funds"
+        );
+    }
+
+    function modifyRailPayment(
+        uint256 railId,
+        uint256 newRate,
+        uint256 oneTimePayment
+    )
+        external
+        validateRailActive(railId)
+        onlyRailOperator(railId)
+        noRailModificationInProgress(railId)
+    {
+        Rail storage rail = rails[railId];
+        Account storage payer = accounts[rail.token][rail.from];
+        Account storage payee = accounts[rail.token][rail.to];
+
+        OperatorApproval storage approval = operatorApprovals[rail.token][
+            rail.from
+        ][rail.operator];
+        require(approval.isApproved, "Operator not approved");
+
+        // Settle the payer's lockup.
+        // If we're changing the rate, we require full settlement.
+        (bool fullySettled, uint256 lockupSettledUpto) = settleAccountLockup(
+            payer
+        );
+        if (newRate != rail.paymentRate) {
+            require(
+                fullySettled && lockupSettledUpto == block.number,
+                "account lockup not fully settled; cannot change rate"
+            );
+        }
+
+        uint256 oldRate = rail.paymentRate;
+        // --- Operator Approval Checks ---
+        validateAndModifyRateChangeApproval(
+            rail,
+            approval,
+            oldRate,
+            newRate,
+            oneTimePayment
+        );
+
+        // --- Settlement Prior to Rate Change ---
+        // If there is no arbiter, settle the rail immediately.
+        if (rail.arbiter == address(0)) {
+            (, uint256 settledUpto, ) = settleRail(railId, block.number);
+            require(
+                settledUpto == block.number,
+                "not able to settle rail upto current epoch"
+            );
+        } else {
+            rail.rateChangeQueue.enqueue(rail.paymentRate, block.number);
```

Comment:
We should probably skip all of this if the rate hasn't changed at all (if we're _just_ making a one-time payment. But it should be a no-op otherwise so I'm happy doing this in a new PR.

Reply:
Doing this in the next PR.